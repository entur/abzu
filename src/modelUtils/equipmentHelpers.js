/*
 *  Licensed under the EUPL, Version 1.2 or â€“ as soon they will be approved by
the European Commission - subsequent versions of the EUPL (the "Licence");
You may not use this work except in compliance with the Licence.
You may obtain a copy of the Licence at:

  https://joinup.ec.europa.eu/software/page/eupl

Unless required by applicable law or agreed to in writing, software
distributed under the Licence is distributed on an "AS IS" basis,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the Licence for the specific language governing permissions and
limitations under the Licence. */

import { defaultEquipmentFacilities, Equipment } from "../models/Equipments";
import { getIn } from "../utils";

const EquipmentHelpers = {};

EquipmentHelpers.isTicketMachinePresent = (entity) => {
  return !!getIn(
    entity,
    ["placeEquipments", "ticketingEquipment", "ticketMachines"],
    null,
  );
};

EquipmentHelpers.isTicketOfficePresent = (entity) => {
  return !!getIn(
    entity,
    ["placeEquipments", "ticketingEquipment", "ticketOffice"],
    null,
  );
};

EquipmentHelpers.getTicketingEquipment = (entity) => {
  return getIn(entity, ["placeEquipments", "ticketingEquipment"], null);
};

EquipmentHelpers.isShelterEquipmentPresent = (entity) => {
  return !!getIn(
    entity,
    ["placeEquipments", "shelterEquipment", "seats"],
    null,
  );
};

EquipmentHelpers.isSanitaryEquipmentPresent = (entity) => {
  return !!getIn(
    entity,
    ["placeEquipments", "sanitaryEquipment", "numberOfToilets"],
    null,
  );
};

EquipmentHelpers.is512SignEquipmentPresent = (entity) => {
  const generalSign = getIn(entity, ["placeEquipments", "generalSign"], null);
  if (
    generalSign &&
    generalSign.privateCode &&
    `${generalSign.privateCode.value}` === "512" &&
    generalSign.signContentType === "transportMode"
  ) {
    return true;
  }
  return false;
};

EquipmentHelpers.update512SignEquipment = (entity, payload) => {
  const copyOfEntity = JSON.parse(JSON.stringify(entity));
  return updateEquipmentForEntity(
    copyOfEntity,
    payload,
    Equipment.GENERAL_SIGN,
  );
};

EquipmentHelpers.isWaitingRoomPresent = (entity) => {
  return !!getIn(
    entity,
    ["placeEquipments", "waitingRoomEquipment", "seats"],
    null,
  );
};

EquipmentHelpers.isCycleStorageEquipmentPresent = (entity) => {
  return !!getIn(
    entity,
    ["placeEquipments", "cycleStorageEquipment", "numberOfSpaces"],
    null,
  );
};

EquipmentHelpers.updateTicketMachineState = (stopPlace, payload) => {
  let updatedStop = JSON.parse(JSON.stringify(stopPlace));
  return updateEquipmentForEntity(
    updatedStop,
    payload,
    Equipment.TICKETING_EQUIPMENT,
  );
};

EquipmentHelpers.updateTicketOfficeState = (stopPlace, payload) => {
  let updatedStop = JSON.parse(JSON.stringify(stopPlace));
  return updateEquipmentForEntity(
    updatedStop,
    payload,
    Equipment.TICKETING_EQUIPMENT,
  );
};

EquipmentHelpers.updateShelterEquipmentState = (stopPlace, payload) => {
  let updatedStop = JSON.parse(JSON.stringify(stopPlace));
  return updateEquipmentForEntity(
    updatedStop,
    payload,
    Equipment.SHELTER_EQUIPMENT,
  );
};

EquipmentHelpers.updateSanitaryEquipmentState = (stopPlace, payload) => {
  let updatedStop = JSON.parse(JSON.stringify(stopPlace));
  return updateEquipmentForEntity(
    updatedStop,
    payload,
    Equipment.SANITARY_EQUIPMENT,
  );
};

EquipmentHelpers.updateWaitingRoomState = (stopPlace, payload) => {
  let updatedStop = JSON.parse(JSON.stringify(stopPlace));
  return updateEquipmentForEntity(
    updatedStop,
    payload,
    Equipment.WAITING_ROOM_EQUIPMENT,
  );
};

EquipmentHelpers.updateCycleStorageEquipmentState = (stopPlace, payload) => {
  let updatedStop = JSON.parse(JSON.stringify(stopPlace));
  return updateEquipmentForEntity(
    updatedStop,
    payload,
    Equipment.CYCLE_STORAGE_EQUIPMENT,
  );
};

const updateEquipmentForEntity = (entity, payload, typeOfEquipment) => {
  const { state, type, id } = payload;

  let stateFromCheckbox = typeof state === "boolean";

  let overrideState = null;

  if (stateFromCheckbox) {
    if (state) {
      overrideState = defaultEquipmentFacilities[typeOfEquipment].isChecked;
    } else {
      overrideState = defaultEquipmentFacilities[typeOfEquipment].isUnChecked;
    }
  } else {
    overrideState = state;
  }

  if (type === "stopPlace") {
    if (!entity.placeEquipments) {
      entity.placeEquipments = {};
    }

    entity.placeEquipments[typeOfEquipment] = overrideState;
  } else if (type === "quay") {
    if (entity.quays && entity.quays[id]) {
      if (!entity.quays[id].placeEquipments) {
        entity.quays[id].placeEquipments = {};
      }
      entity.quays[id].placeEquipments[typeOfEquipment] = overrideState;
    }
  }

  return entity;
};

export default EquipmentHelpers;
