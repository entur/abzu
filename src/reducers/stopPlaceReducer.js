/*
 *  Licensed under the EUPL, Version 1.2 or â€“ as soon they will be approved by
the European Commission - subsequent versions of the EUPL (the "Licence");
You may not use this work except in compliance with the Licence.
You may obtain a copy of the Licence at:

  https://joinup.ec.europa.eu/software/page/eupl

Unless required by applicable law or agreed to in writing, software
distributed under the Licence is distributed on an "AS IS" basis,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the Licence for the specific language governing permissions and
limitations under the Licence. */

import * as types from "../actions/Types";
import AdjacentStopAdder from "../modelUtils/adjacentStopAdder";
import AdjacentStopRemover from "../modelUtils/adjacentStopRemover";
import equipmentHelpers from "../modelUtils/equipmentHelpers";
import limitationHelpers from "../modelUtils/limitationHelpers";
import formatHelpers from "../modelUtils/mapToClient";
import SettingsManager from "../singletons/SettingsManager";
import { setDecimalPrecision } from "../utils/";
import { getStateByOperation } from "./stopPlaceReducerUtils";

const Settings = new SettingsManager();

const initialState = {
  centerPosition: [64.349421, 16.809082],
  zoom: 6,
  minZoom: 14,
  isCompassBearingEnabled: Settings.getShowCompassBearing(),
  isCreatingPolylines: false,
  enablePublicCodePrivateCodeOnStopPlaces:
    Settings.getEnablePublicCodePrivateCodeOnStopPlaces(),
  enablePolylines: Settings.getShowPathLinks(),
  showExpiredStops: Settings.getShowExpiredStops(),
  showMultimodalEdges: Settings.getShowMultimodalEdges(),
  lastMutatedStopPlaceId: [],
  isFetchingMergeInfo: false,
  loading: false,
};

const stopPlaceReducer = (state = initialState, action) => {
  switch (action.type) {
    /* These actions are dispatched by Apollo-Client */
    case types.APOLLO_QUERY_RESULT:
    case types.APOLLO_MUTATION_RESULT:
      return getStateByOperation(state, action);

    case types.CLEAR_SEARCH_RESULTS:
      return Object.assign({}, state, {
        searchResults: [],
      });

    case types.REMOVED_CHILD_FROM_PARENT_STOP_PLACE:
      return Object.assign({}, state, {
        current: formatHelpers.updateParenStopWithoutStopPlace(
          state.current,
          action.payload,
        ),
        stopHasBeenModified: true,
      });

    case types.ADDED_STOP_PLACES_TO_PARENT:
      return Object.assign({}, state, {
        current: formatHelpers.updateParentStopWithStopPlaces(
          state.current,
          action.payload,
        ),
        stopHasBeenModified: true,
      });

    case types.DESTROYED_NEW_STOP:
      return Object.assign({}, state, {
        newStop: null,
      });

    case types.NAVIGATE_TO_MAIN_AFTER_DELETE:
      return Object.assign({}, state, {
        searchResults: [],
        pathLink: [],
        current: null,
        stopHasBeenModified: false,
        activeSearchResult: null,
        versions: [],
        originalCurrent: null,
      });

    case types.HID_QUAYS_FOR_NEIGHBOUR_STOP:
      const id = action.payload;
      return Object.assign({}, state, {
        neighbourStopQuays: { ...state.neighbourStopQuays, [id]: null },
      });

    case types.RESTORED_TO_ORIGINAL_STOP_PLACE:
      return Object.assign({}, state, {
        stopHasBeenModified: false,
        current: JSON.parse(JSON.stringify(state.originalCurrent)),
        pathLink: JSON.parse(JSON.stringify(state.originalPathLink)),
      });

    case types.ADD_ADJACENT_SITE:
      const stopPlaceId1 = action.payload.stopPlaceId1;
      const stopPlaceId2 = action.payload.stopPlaceId2;
      AdjacentStopAdder.addAdjacentStopReference(
        state.current,
        stopPlaceId1,
        stopPlaceId2,
      );

      return Object.assign({}, state, {
        stopHasBeenModified: true,
      });

    case types.REMOVE_ADJACENT_SITE:
      const adjacentStopPlaceRef = action.payload.adjacentStopPlaceRef;
      const stopPlaceIdForRemovingAdjacentSite = action.payload.stopPlaceId;
      const changedStopPlace = AdjacentStopRemover.removeAdjacentStop(
        state.current,
        adjacentStopPlaceRef,
        stopPlaceIdForRemovingAdjacentSite,
      );

      return Object.assign({}, state, {
        stopHasBeenModified: true,
        current: changedStopPlace,
      });

    case types.SET_CENTER_AND_ZOOM:
      return Object.assign({}, state, {
        centerPosition: action.payload.position.slice(),
        zoom: action.payload.zoom,
      });

    case types.CLEAR_LAST_MUTATED_STOP_PLACE_IDS:
      return Object.assign({}, state, {
        lastMutatedStopPlaceId: [],
      });

    case types.NAVIGATE_TO:
      if (action.payload === "") {
        return Object.assign({}, state, {
          pathLink: [],
          current: null,
          newStop: null,
        });
      } else {
        return state;
      }

    case types.REMOVED_STOPS_NEARBY_FOR_OVERVIEW:
      return Object.assign({}, state, {
        neighbourStops: [],
      });

    case types.SORTED_QUAYS:
      return Object.assign({}, state, {
        current: formatHelpers.sortQuays(state.current, action.payload),
      });

    case types.LOOKUP_COORDINATES:
      return Object.assign({}, state, {
        findCoordinates: {
          position: action.payload.position.map((pos) =>
            setDecimalPrecision(pos, 6),
          ),
          coordinatePin: true,
        },
        centerPosition: action.payload.triggeredByDrag
          ? state.centerPosition
          : action.payload.position,
        zoom: action.payload.triggeredByDrag ? state.zoom : 5,
      });

    case types.CHANGED_WEIGHTING_STOP_PLACE:
      return Object.assign({}, state, {
        current: formatHelpers.updateCurrentStopWithWeighting(
          state.current,
          action.payload,
        ),
        stopHasBeenModified: true,
      });

    case types.CREATED_NEW_STOP:
      const { location, isMultimodal, locationPermissions } = action.payload;
      const stopToBeCreated = isMultimodal
        ? formatHelpers.createNewParentStopFromLocation(location)
        : formatHelpers.createNewStopFromLocation(location);

      return Object.assign({}, state, {
        newStop: {
          ...stopToBeCreated,
          permissions: locationPermissions,
        },
        originalCurrent: JSON.parse(JSON.stringify(stopToBeCreated)),
        versions: [],
        pathLink: [],
        newStopIsMultiModal: isMultimodal,
        stopHasBeenModified: false,
      });

    case types.CHANGED_QUAY_COMPASS_BEARING:
      return Object.assign({}, state, {
        current: formatHelpers.updateCompassBearing(
          state.current,
          action.payload,
        ),
        stopHasBeenModified: true,
      });

    case types.UPDATED_KEY_VALUES_FOR_KEY:
      return Object.assign({}, state, {
        current: formatHelpers.updateKeyValuesByKey(
          state.current,
          action.payload.key,
          action.payload.values,
          action.payload.origin,
        ),
        stopHasBeenModified: true,
      });

    case types.DELETED_KEY_VALUES_BY_KEY:
      return Object.assign({}, state, {
        current: formatHelpers.deleteKeyValuesByKey(
          state.current,
          action.payload.key,
          action.payload.origin,
        ),
        stopHasBeenModified: true,
      });

    case types.CREATED_KEY_VALUES_PAIR:
      return Object.assign({}, state, {
        current: formatHelpers.createKeyValuesPair(
          state.current,
          action.payload.key,
          action.payload.values,
          action.payload.origin,
        ),
        stopHasBeenModified: true,
      });

    case types.CHANGED_LOCATION_NEW_STOP:
      return Object.assign({}, state, {
        newStop: formatHelpers.createNewStopFromLocation(action.payload),
        stopHasBeenModified: true,
      });

    case types.CHANGED_STOP_NAME:
      return {
        ...state,
        current: {
          ...state.current,
          name: action.payload,
        },
        stopHasBeenModified: true,
      };

    case types.CHANGED_STOP_PUBLIC_CODE:
      return {
        ...state,
        current: {
          ...state.current,
          publicCode: action.payload,
        },
        stopHasBeenModified: true,
      };

    case types.CHANGED_STOP_PRIVATE_CODE:
      return {
        ...state,
        current: {
          ...state.current,
          privateCode: action.payload,
        },
        stopHasBeenModified: true,
      };

    case types.CHANGED_STOP_DESCRIPTION:
      return {
        ...state,
        current: {
          ...state.current,
          description: action.payload,
        },
        stopHasBeenModified: true,
      };

    case types.CHANGED_STOP_TYPE:
      return Object.assign({}, state, {
        current: formatHelpers.updateCurrentStopWithType(
          state.current,
          action.payload,
        ),
        stopHasBeenModified: true,
      });

    case types.CHANGED_STOP_SUBMODE:
      return Object.assign({}, state, {
        current: formatHelpers.updateCurrentStopWithSubMode(
          state.current,
          action.payload.stopPlaceType,
          action.payload.transportMode,
          action.payload.submode,
        ),
        stopHasBeenModified: true,
      });

    case types.CHANGED_ACTIVE_STOP_POSITION:
      return Object.assign({}, state, {
        current: formatHelpers.updateCurrentStopWithPosition(
          state.current,
          action.payload.location,
        ),
        stopHasBeenModified: true,
      });

    case types.USE_NEW_STOP_AS_CURRENT:
      return Object.assign({}, state, {
        current: JSON.parse(JSON.stringify(state.newStop)),
        centerPosition: state.newStop.location,
        isCreatingPolylines: false,
        zoom: 14,
        stopHasBeenModified: false,
        versions: [],
      });

    case types.CREATE_NEW_MULTIMODAL_STOP_FROM_EXISTING:
      const { newStopPlace, fromMain } = action.payload;
      let newState = Object.assign({}, state, {
        current: newStopPlace,
        isCreatingPolylines: false,
        stopHasBeenModified: true,
        versions: [],
      });
      if (!fromMain) {
        newState.centerPosition = newStopPlace.location;
      }
      return newState;

    case types.SET_ACTIVE_MARKER:
      return Object.assign({}, state, {
        activeSearchResult: action.payload,
        centerPosition: getProperCenterLocation(action.payload.location),
        zoom: getProperZoomLevel(action.payload.location),
      });

    case types.SET_MISSING_COORDINATES:
      return Object.assign({}, state, {
        centerPosition: getProperCenterLocation(action.payload.position),
        zoom: getProperZoomLevel(action.payload.position),
        userDefinedCoordinates: action.payload,
        activeSearchResult: Object.assign({}, state.activeSearchResult, {
          location: action.payload.position,
        }),
      });

    case types.ADDED_STOP_PLACE_ELEMENT:
      return Object.assign({}, state, {
        current: formatHelpers.updateCurrentWithNewElement(
          state.current,
          action.payload,
        ),
        stopHasBeenModified: true,
      });

    case types.CHANGE_ELEMENT_POSITION:
      return Object.assign({}, state, {
        current: formatHelpers.updateCurrentWithElementPositionChange(
          state.current,
          action.payload,
        ),
        stopHasBeenModified: true,
      });

    case types.CHANGE_PUBLIC_CODE_NAME:
      return Object.assign({}, state, {
        current: formatHelpers.updateCurrentWithPublicCode(
          state.current,
          action.payload,
        ),
        stopHasBeenModified: true,
      });

    case types.CHANGE_PRIVATE_CODE_NAME:
      return Object.assign({}, state, {
        current: formatHelpers.updateCurrentWithPrivateCode(
          state.current,
          action.payload,
        ),
        stopHasBeenModified: true,
      });

    case types.CHANGED_ELEMENT_DESCRIPTION:
      return Object.assign({}, state, {
        current: formatHelpers.updateCurrentWithElementDescriptionChange(
          state.current,
          action.payload,
        ),
        stopHasBeenModified: true,
      });

    case types.REMOVED_ELEMENT_BY_TYPE:
      return Object.assign({}, state, {
        current: formatHelpers.updateCurrentWithoutElement(
          state.current,
          action.payload,
        ),
        stopHasBeenModified: true,
      });

    case types.CHANGED_MAP_CENTER:
      return Object.assign({}, state, {
        centerPosition: action.payload.position,
        zoom: action.payload.zoom,
      });

    case types.STARTED_CREATING_POLYLINE:
      return Object.assign({}, state, {
        pathLink: formatHelpers.updatePathLinkWithNewEntry(
          action,
          state.pathLink,
        ),
        stopHasBeenModified: true,
        isCreatingPolylines: true,
        enablePolylines: true,
      });

    case types.ADDED_FINAL_COORDINATES_TO_POLYLINE:
      return Object.assign({}, state, {
        pathLink: formatHelpers.updatePathLinkWithNewEntry(
          action,
          state.pathLink,
        ),
        stopHasBeenModified: true,
        isCreatingPolylines: false,
      });

    case types.ADDED_COORDINATES_TO_POLYLINE:
      return Object.assign({}, state, {
        pathLink: formatHelpers.addNewPointToPathlink(action, state.pathLink),
        enablePolylines: true,
      });

    case types.REMOVED_LAST_POLYLINE:
      return Object.assign({}, state, {
        pathLink: state.pathLink.slice(0, state.pathLink.length - 1),
        isCreatingPolylines: false,
      });

    case types.EDITED_TIME_ESTIMATE_FOR_POLYLINE:
      return Object.assign({}, state, {
        pathLink: formatHelpers.updateEstimateForPathLink(
          action,
          state.pathLink,
        ),
      });

    case types.TOGGLED_ENABLE_PUBLIC_CODE_PRIVATE_CODE_ON_STOP_PLACES:
      return Object.assign({}, state, {
        enablePublicCodePrivateCodeOnStopPlaces: action.payload,
      });

    case types.TOGGLED_IS_MULTIPOLYLINES_ENABLED:
      return Object.assign({}, state, { enablePolylines: action.payload });

    case types.TOGGLED_IS_COMPASS_BEARING_ENABLED:
      return Object.assign({}, state, {
        isCompassBearingEnabled: action.payload,
      });

    case types.TOGGLED_IS_MULTIMODAL_EDGES_ENABLED:
      return Object.assign({}, state, {
        showMultimodalEdges: action.payload,
      });

    case types.CHANGED_STOP_ACCESSIBLITY_ASSESSMENT:
      return Object.assign({}, state, {
        current: limitationHelpers.updateCurrentWithLimitations(
          state.current,
          action.payload,
        ),
        stopHasBeenModified: true,
      });

    case types.CHANGED_QUAY_ACCESSIBLITY_ASSESSMENT:
      return Object.assign({}, state, {
        current: limitationHelpers.updateCurrentWithQuayLimitations(
          state.current,
          action.payload,
        ),
        stopHasBeenModified: true,
      });

    case types.CHANGED_TICKET_MACHINE_STATE:
      return Object.assign({}, state, {
        current: equipmentHelpers.updateTicketMachineState(
          state.current,
          action.payload,
        ),
        stopHasBeenModified: true,
      });

    case types.CHANGED_SHELTER_EQUIPMENT_STATE:
      return Object.assign({}, state, {
        current: equipmentHelpers.updateShelterEquipmentState(
          state.current,
          action.payload,
        ),
        stopHasBeenModified: true,
      });

    case types.CHANGED_SANITARY_EQUIPMENT_STATE:
      return Object.assign({}, state, {
        current: equipmentHelpers.updateSanitaryEquipmentState(
          state.current,
          action.payload,
        ),
        stopHasBeenModified: true,
      });

    case types.CHANGED_WAITING_ROOM_STATE:
      return Object.assign({}, state, {
        current: equipmentHelpers.updateWaitingRoomState(
          state.current,
          action.payload,
        ),
        stopHasBeenModified: true,
      });

    case types.CHANGED_CYCLE_STORAGE_STATE:
      return Object.assign({}, state, {
        current: equipmentHelpers.updateCycleStorageEquipmentState(
          state.current,
          action.payload,
        ),
        stopHasBeenModified: true,
      });

    case types.CHANGED_TRANSPORT_SIGN_STATE:
      return Object.assign({}, state, {
        current: equipmentHelpers.update512SignEquipment(
          state.current,
          action.payload,
        ),
        stopHasBeenModified: true,
      });

    case types.ADDED_ALT_NAME:
      return Object.assign({}, state, {
        current: formatHelpers.addAltName(state.current, action.payload),
        stopHasBeenModified: true,
      });

    case types.EDITED_ALT_NAME:
      return Object.assign({}, state, {
        current: formatHelpers.editAltName(state.current, action.payload),
        stopHasBeenModified: true,
      });

    case types.REMOVED_ALT_NAME:
      return Object.assign({}, state, {
        current: formatHelpers.removeAltName(state.current, action.payload),
        stopHasBeenModified: true,
      });

    case types.CHANGED_PARKING_NAME:
      return Object.assign({}, state, {
        current: formatHelpers.changeParkingName(state.current, action.payload),
        stopHasBeenModified: true,
      });

    case types.CHANGED_PARKING_LAYOUT:
      return Object.assign({}, state, {
        current: formatHelpers.changeParkingLayout(
          state.current,
          action.payload,
        ),
        stopHasBeenModified: true,
      });

    case types.CHANGED_PARKING_PAYMENT_PROCESS:
      return Object.assign({}, state, {
        current: formatHelpers.changeParkingPaymentProcess(
          state.current,
          action.payload,
        ),
        stopHasBeenModified: true,
      });

    case types.CHANGED_PARKING_RECHARGING_AVAILABLE:
      return Object.assign({}, state, {
        current: formatHelpers.changeParkingRechargingAvailable(
          state.current,
          action.payload,
        ),
        stopHasBeenModified: true,
      });

    case types.CHANGED_PARKING_NUMBER_OF_SPACES:
      return Object.assign({}, state, {
        current: formatHelpers.changeParkingNumberOfSpaces(
          state.current,
          action.payload,
        ),
        stopHasBeenModified: true,
      });

    case types.CHANGED_PARKING_NUMBER_OF_SPACES_WITH_RECHARGE_POINT:
      return Object.assign({}, state, {
        current: formatHelpers.changeParkingNumberOfSpacesWithRechargePoint(
          state.current,
          action.payload,
        ),
        stopHasBeenModified: true,
      });

    case types.CHANGED_PARKING_NUMBER_OF_SPACES_FOR_REGISTERED_DISABLED_USER_TYPE:
      return Object.assign({}, state, {
        current:
          formatHelpers.changeParkingNumberOfSpacesForRegisteredDisabledUserType(
            state.current,
            action.payload,
          ),
        stopHasBeenModified: true,
      });

    case types.CHANGED_PARKING_TOTAL_CAPACITY:
      return Object.assign({}, state, {
        current: formatHelpers.changeParkingTotalCapacity(
          state.current,
          action.payload,
        ),
        stopHasBeenModified: true,
      });

    case types.OPENED_MERGE_STOP_DIALOG:
      return Object.assign({}, state, {
        mergeStopDialog: {
          isOpen: true,
          ...action.payload,
        },
      });

    case types.REQUESTED_QUAYS_MERGE_INFO:
      return Object.assign({}, state, {
        isFetchingMergeInfo: true,
      });

    case types.RECEIVED_QUAYS_MERGE_INFO:
      return Object.assign({}, state, {
        isFetchingMergeInfo: false,
      });

    case types.CLOSED_MERGE_STOP_DIALOG:
      return Object.assign({}, state, {
        mergeStopDialog: {
          isOpen: false,
        },
      });

    case types.TOGGLED_IS_SHOW_EXPIRED_STOPS:
      return Object.assign({}, state, {
        showExpiredStops: action.payload,
      });

    case types.SET_STOP_PLACE_LOADING:
      return Object.assign({}, state, {
        loading: action.payload,
      });

    default:
      return state;
  }
};

const getProperCenterLocation = (location) => {
  return location || [62.928595, 12.083002];
};

const getProperZoomLevel = (location) => {
  return location ? 15 : 5;
};

export default stopPlaceReducer;
