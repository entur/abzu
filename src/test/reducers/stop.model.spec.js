/*
 *  Licensed under the EUPL, Version 1.2 or â€“ as soon they will be approved by
the European Commission - subsequent versions of the EUPL (the "Licence");
You may not use this work except in compliance with the Licence.
You may obtain a copy of the Licence at:

  https://joinup.ec.europa.eu/software/page/eupl

Unless required by applicable law or agreed to in writing, software
distributed under the Licence is distributed on an "AS IS" basis,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the Licence for the specific language governing permissions and
limitations under the Licence. */

import stopPlaceReducer from "./../../reducers/stopPlaceReducer";
import stopPlaceMock from "./json/stopPlace.json";
import stopPlaceMock10Quays from "./json/stopPlaceWith10Quays.json";
import clientStop from "./json/clientStop.json";
import QueryVariablesMapper from "../../modelUtils/mapToQueryVariables";
import { hasExpired } from "../../modelUtils/validBetween";

window.config = {
  defaultLanguageCode: "nor",
};

describe("Model: map format from server to expected client model", () => {
  test("should map GraphQL response to client model for StopPlace", () => {
    const action = {
      type: "APOLLO_QUERY_RESULT",
      result: stopPlaceMock,
      operationName: "stopPlace",
    };
    const state = stopPlaceReducer({}, action);

    expect(state.current).toMatchSnapshot();
  });

  test("should map parking to schema validated query variables", () => {
    const parking = [
      {
        name: "park&ride example",
        location: [63.207698, 11.088595],
        totalCapacity: "100",
        parkingVehicleTypes: ["car"],
        validBetween: null,
      },
    ];

    let result = QueryVariablesMapper.mapParkingToVariables(
      parking,
      "NSR:StopPlace:1",
    );

    expect(result).toMatchSnapshot();
  });

  test("should map client stop to schema correctly", () => {
    const schemaValidStop = QueryVariablesMapper.mapStopToVariables(clientStop);
    expect(schemaValidStop).toMatchSnapshot();
  });
});

describe("Changes correct properties", () => {
  var state = {};

  beforeAll((done) => {
    const action = {
      type: "APOLLO_QUERY_RESULT",
      result: stopPlaceMock10Quays,
      operationName: "stopPlace",
    };
    state = stopPlaceReducer({}, action);
    expect(state.current.quays.length).toEqual(10);
    done();
  });

  test("should change property of correct quay", () => {
    for (let quayIndex = 0; quayIndex < 10; quayIndex++) {
      const newPublicCode = `new public code ${quayIndex}`;

      const changePublicCode = {
        type: "CHANGE_PUBLIC_CODE_NAME",
        payload: {
          type: "quay",
          name: newPublicCode,
          index: quayIndex,
        },
      };

      state = stopPlaceReducer(state, changePublicCode);

      expect(state.current.quays[quayIndex].publicCode).toEqual(newPublicCode);

      const stopValidWithSchema = QueryVariablesMapper.mapStopToVariables(
        state.current,
      );

      expect(state.current.quays[quayIndex].id).toEqual(
        stopValidWithSchema.quays[quayIndex].id,
      );
    }
  });

  test("Should correctly determine if a stop has expired or not based on validBetween", () => {
    const expiredDate = {
      fromDate: "2017-05-31T11:03:01.770+0200",
      toDate: "2017-05-31T11:03:01.842+0200",
    };

    const hasStopExpired = hasExpired(expiredDate);

    expect(hasStopExpired).toEqual(true);

    const validDate = {
      fromDate: "2017-05-31T11:03:01.770+0200",
      toDate: null,
    };

    const hasStopExpired2 = hasExpired(validDate);

    expect(hasStopExpired2).toEqual(false);
  });
});
