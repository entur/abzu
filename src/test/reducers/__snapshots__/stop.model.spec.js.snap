// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Model: map format from server to expected client model should map GraphQL response to client model for StopPlace 1`] = `
Object {
  "accessibilityAssessment": Object {
    "limitations": Object {
      "audibleSignalsAvailable": "UNKNOWN",
      "escalatorFreeAccess": "UNKNOWN",
      "liftFreeAccess": "UNKNOWN",
      "stepFreeAccess": "UNKNOWN",
      "wheelchairAccess": "UNKNOWN",
    },
  },
  "adjacentSites": undefined,
  "alternativeNames": Array [],
  "belongsToGroup": false,
  "entityType": "STOP_PLACE",
  "fareZones": Array [],
  "groups": Array [],
  "hasExpired": false,
  "id": "NSR:StopPlace:933",
  "importedId": Array [],
  "isActive": true,
  "isFuture": false,
  "keyValues": Array [],
  "location": Array [
    60.260427,
    5.435734,
  ],
  "name": "Aspestrand",
  "parentTopographicPlace": "Ã˜stfold",
  "parking": Array [],
  "permanentlyTerminated": false,
  "quays": Array [
    Object {
      "accessibilityAssessment": Object {
        "limitations": Object {
          "audibleSignalsAvailable": "UNKNOWN",
          "escalatorFreeAccess": "UNKNOWN",
          "liftFreeAccess": "UNKNOWN",
          "stepFreeAccess": "UNKNOWN",
          "wheelchairAccess": "UNKNOWN",
        },
      },
      "boardingPositions": Array [],
      "compassBearing": 344,
      "description": "",
      "id": "NSR:Quay:1694",
      "importedId": Array [],
      "keyValues": Array [],
      "location": Array [
        60.260427,
        5.435734,
      ],
      "publicCode": "1",
    },
    Object {
      "accessibilityAssessment": Object {
        "limitations": Object {
          "audibleSignalsAvailable": "UNKNOWN",
          "escalatorFreeAccess": "UNKNOWN",
          "liftFreeAccess": "UNKNOWN",
          "stepFreeAccess": "UNKNOWN",
          "wheelchairAccess": "UNKNOWN",
        },
      },
      "boardingPositions": Array [],
      "compassBearing": 164,
      "description": "",
      "id": "NSR:Quay:1695",
      "importedId": Array [],
      "keyValues": Array [],
      "location": Array [
        60.260427,
        5.435734,
      ],
      "publicCode": "2",
    },
  ],
  "stopPlaceType": "onstreetBus",
  "submode": null,
  "tags": Array [],
  "tariffZones": Array [],
  "topographicPlace": "Aremark",
  "transportMode": null,
  "validBetween": Object {
    "fromDate": "2017-05-26T14:52:10.096+0200",
    "toDate": null,
  },
  "version": 1,
  "weighting": null,
}
`;

exports[`Model: map format from server to expected client model should map client stop to schema correctly 1`] = `
Object {
  "accessibilityAssessment": Object {
    "limitations": Object {
      "audibleSignalsAvailable": "UNKNOWN",
      "escalatorFreeAccess": "UNKNOWN",
      "liftFreeAccess": "UNKNOWN",
      "stepFreeAccess": "UNKNOWN",
      "wheelchairAccess": "UNKNOWN",
    },
  },
  "adjacentSites": undefined,
  "alternativeNames": Array [],
  "description": Object {
    "lang": "nor",
    "value": "Beskrivelse",
  },
  "id": "NSR:StopPlace:19744",
  "keyValues": Array [],
  "legacyCoordinates": Array [
    Array [
      11.170963,
      59.587427,
    ],
  ],
  "name": Object {
    "lang": "nor",
    "value": "Aspelundsveien",
  },
  "placeEquipments": null,
  "privateCode": Object {
    "value": "private",
  },
  "publicCode": "public",
  "quays": Array [
    Object {
      "accessibilityAssessment": undefined,
      "compassBearing": 212,
      "description": null,
      "geometry": Object {
        "legacyCoordinates": Array [
          Array [
            11.17094,
            59.587486,
          ],
        ],
        "type": "Point",
      },
      "id": "NSR:Quay:30025",
      "keyValues": Array [],
      "placeEquipments": null,
      "privateCode": Object {
        "value": "",
      },
      "publicCode": null,
    },
    Object {
      "accessibilityAssessment": undefined,
      "compassBearing": 38,
      "description": null,
      "geometry": Object {
        "legacyCoordinates": Array [
          Array [
            11.170986,
            59.587368,
          ],
        ],
        "type": "Point",
      },
      "id": "NSR:Quay:30026",
      "keyValues": Array [],
      "placeEquipments": null,
      "privateCode": Object {
        "value": "",
      },
      "publicCode": null,
    },
  ],
  "stopPlaceType": "onstreetBus",
  "submode": null,
  "tariffZones": Array [],
  "transportMode": null,
  "weighting": null,
}
`;

exports[`Model: map format from server to expected client model should map parking to schema validated query variables 1`] = `
Array [
  Object {
    "geometry": Object {
      "legacyCoordinates": Array [
        Array [
          11.088595,
          63.207698,
        ],
      ],
      "type": "Point",
    },
    "name": Object {
      "lang": "nor",
      "value": "park&ride example",
    },
    "parentSiteRef": "NSR:StopPlace:1",
    "parkingVehicleTypes": Array [
      "car",
    ],
    "totalCapacity": 100,
    "validBetween": null,
  },
]
`;
